packages/actor-http-limit-rate/lib/ActorHttpLimitRate.ts

```diff
@@ -0,0 +1,212 @@
+import type { IActionHttp, IActorHttpOutput, IActorHttpArgs, MediatorHttp } from '@comunica/bus-http';
+import { ActorHttp } from '@comunica/bus-http';
+import type { ActorHttpInvalidateListenable, IActionHttpInvalidate } from '@comunica/bus-http-invalidate';
+import type { TestResult } from '@comunica/core';
+import { ActionContextKey, failTest, passTest } from '@comunica/core';
+import type { IMediatorTypeTime } from '@comunica/mediatortype-time';
+
+export class ActorHttpLimitRate extends ActorHttp {
+  private readonly historyLength: number;
+  private readonly failureMultiplier: number;
+  private readonly hostData: Record<string, IHostRequestData>;
+  private readonly httpInvalidator: ActorHttpInvalidateListenable;
+  private readonly mediatorHttp: MediatorHttp;
+
+  // Context key to indicate that the actor has already wrapped the given request
+  private static readonly keyWrapped = new ActionContextKey<boolean>('urn:comunica:actor-http-limit-rate#wrapped');
+
+  public constructor(args: IActorHttpLimitConcurrentArgs) {
+    super(args);
+    this.mediatorHttp = args.mediatorHttp;
+    this.httpInvalidator = args.httpInvalidator;
+    this.httpInvalidator.addInvalidateListener(action => this.handleHttpInvalidateEvent(action));
+    this.historyLength = args.historyLength;
+    this.failureMultiplier = args.failureMultiplier;
+    this.hostData = {};
+  }
+
+  public async test(action: IActionHttp): Promise<TestResult<IMediatorTypeTime>> {
+    if (action.context.has(ActorHttpLimitRate.keyWrapped)) {
+      return failTest(`${this.name} can only wrap a request once`);
+    }
+    return passTest({ time: 0 });
+  }
+
+  public async run(action: IActionHttp): Promise<IActorHttpOutput> {
+    const host = ActorHttp.getInputUrl(action.input).host;
+
+    // Wait for the next free request slot on the host
+    await this.waitForSlot(host);
+
+    // Delay the request when relevant
+    await this.delayRequest(host);
+
+    const timeStart = Date.now();
+
+    this.hostData[host].latestRequestTimestamp = timeStart;
+
+    const response = await this.mediatorHttp.mediate({
```

Comment:
The fetch operation could also be rejected.
I guess this would need to be try-catched here to correctly handle the hostData?